private/protocol/xml/xmlutil/xml_to_struct.go

```diff
@@ -59,41 +63,52 @@ func XMLToStruct(d *xml.Decoder, s *xml.StartElement) (*XMLNode, error) {
 				slice = []*XMLNode{}
 			}
 			node, e := XMLToStruct(d, &el)
+			out.findNamespaces()
 			if e != nil {
 				return out, e
 			}
 			node.Name = typed.Name
-			node.Attr = out.Attr
-			node = adaptNode(node)
+			node.findNamespaces()
+			tempOut := *out
+			// Save into a temp variable, simply because out gets squashed during
+			// loop iterations
+			node.parent = &tempOut
 			slice = append(slice, node)
 			out.Children[name] = slice
 		case xml.EndElement:
 			if s != nil && s.Name.Local == typed.Name.Local { // matching end token
 				return out, nil
 			}
+			out = &XMLNode{}
 		}
 	}
 	return out, nil
 }
 
-func adaptNode(node *XMLNode) *XMLNode {
+func (n *XMLNode) findNamespaces() {
 	ns := map[string]string{}
-	for _, a := range node.Attr {
+	for _, a := range n.Attr {
 		if a.Name.Space == "xmlns" {
 			ns[a.Value] = a.Name.Local
-			break
 		}
 	}
 
-	for i, a := range node.Attr {
-		if a.Name.Space == "xmlns" {
-			continue
-		}
-		if v, ok := ns[node.Attr[i].Name.Space]; ok {
-			node.Attr[i].Name.Space = v
+	n.namespaces = ns
+}
+
+func (n *XMLNode) findElem(name string) (string, bool) {
+	for node := n; node != nil; node = node.parent {
+		for _, a := range node.Attr {
+			namespace := a.Name.Space
+			if v, ok := node.namespaces[namespace]; ok {
+				namespace = v
+			}
+			if name == strings.Join([]string{namespace, a.Name.Local}, ":") {
```

Comment:
Would it make sense to use:

`name == namespace + ":" + a.Name.Local`

or

`name == fmt.Sprintf("%s:%s", namespace, a.Name.Local`)`

Reply:
prefer not using sprintf as that may need to use runtime reflection